src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,395 @@
+            if (parameterWithMatchingName.Ordinal >= arguments.Count)
+            {
+                return null;
+            }
+
+            // Multiple arguments could have been converted to a single params array, 
+            // so there wouldn't be a corresponding argument
+            if (parameterWithMatchingName.IsParams && parameters.Length != arguments.Count)
+            {
+                return null;
+            }
+
+            return arguments[parameterWithMatchingName.Ordinal];
+        }
+
+        private IParameterSymbol GetParameterWithMatchingName(ImmutableArray<IParameterSymbol> parameters, string searchArgumentName)
+        {
+            foreach (var p in parameters)
+            {
+                if (p.Name.Equals(searchArgumentName))
+                {
+                    return p;
+                }
+            }
+
+            return null;
+        }
+
+        protected SyntaxNode TryGetFormatStringLiteralExpressionSyntax(
+            SemanticModel semanticModel,
+            SeparatedSyntaxList<SyntaxNode> arguments,
+            ImmutableArray<IParameterSymbol> parameters)
+        {
+            var formatArgumentSyntax = TryGetArgument(
+                semanticModel,
+                NameOfFormatStringParameter,
+                arguments,
+                parameters);
+
+            if (formatArgumentSyntax == null)
+            {
+                return null;
+            }
+
+            if (!ArgumentExpressionIsStringLiteral(formatArgumentSyntax))
+            {
+                return null;
+            }
+
+            return GetArgumentExpression(formatArgumentSyntax);
+        }
+
+        protected static IMethodSymbol TryGetValidFormatMethodSymbol(
+            SyntaxNodeAnalysisContext context,
+            int numberOfArguments,
+            SymbolInfo symbolInfo)
+        {
+
+            if (symbolInfo.Symbol == null)
+            {
+                return null;
+            }
+
+            if (symbolInfo.Symbol.Kind != SymbolKind.Method)
+            {
+                return null;
+            }
+
+            var containingType = (INamedTypeSymbol)symbolInfo.Symbol.ContainingSymbol;
+
+            if (containingType.SpecialType != SpecialType.System_String)
+            {
+                return null;
+            }
+
+            var method = (IMethodSymbol)symbolInfo.Symbol;
+
+            if (method.Arity > 0)
+            {
+                return null;
+            }
+
+            return method;
+        }
+
+        private bool FormatCallWorksAtRuntime(string formatString, int numberOfPlaceholderArguments)
+        {
+            var testArray = new object[numberOfPlaceholderArguments];
+            for (var i = 0; i < numberOfPlaceholderArguments; i++)
+            {
+                testArray[i] = "test";
+            }
+
+            try
+            {
+                string.Format(formatString, testArray);
+            }
+            catch
```

Comment:
i am so torn on how i feel about this.  @Pilchie @jasonmalinowski, thoughts?  i really don't like that we're making it the norm that our code would throw exceptions expectedly as part of processing user code.